service/collector/elb.go

```diff
@@ -35,27 +34,76 @@ var (
 	)
 )
 
-func (c *Collector) collectAccountsELBs(ch chan<- prometheus.Metric, clients []aws.Clients) {
-	var wg sync.WaitGroup
+type ELBConfig struct {
+	Helper *helper
+	Logger micrologger.Logger
+
+	InstallationName string
+}
+
+type ELBCollector struct {
+	helper *helper
+	logger micrologger.Logger
+
+	installationName string
+}
+
+func NewELB(config ELBConfig) (*ELBCollector, error) {
+	if config.Helper == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Helper must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.InstallationName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.InstallationName must not be empty", config)
+	}
+
+	c := &ELBCollector{
+		helper: config.Helper,
+		logger: config.Logger,
+
+		installationName: config.InstallationName,
+	}
+
+	return c, nil
+}
 
-	for _, client := range clients {
-		wg.Add(1)
-		go func(awsClients aws.Clients) {
-			defer wg.Done()
-			err := c.collectELBs(ch, awsClients)
+func (c *ELBCollector) Collect(ch chan<- prometheus.Metric) error {
+	awsClientsList, err := c.helper.GetAWSClients()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var g errgroup.Group
+
+	for _, awsClients := range awsClientsList {
+		g.Go(func() error {
+			err := c.collectForAccount(ch, awsClients)
 			if err != nil {
-				c.logger.Log("level", "error", "message", "failed collecting ELB metrics", "stack", fmt.Sprintf("%#v", err))
+				return microerror.Mask(err)
```

Comment:
We had all errors logged. Now only the fist one. Do you think it's better?

Reply:
We never had the case but in case an error occurs on this level it is likely to happen again. We retry here anyway due to the scrape interval. We are going to try again anyway. My thinking was that the immediate continuation is just increasing more load and now we behave more like a backoff.

Reply:
That makes sense.